---
title: "Class 13: Transcriptomics & Analysis of RNA-Seq Data"
author: "Katherine Quach (A18541014)"
format: pdf
toc: TRUE
---
## Background

Today we will perform an RNASeq analysis of the effects of a common steroid on airway cells. 

In particular, dexamethasone (just called "dex") on different airway smooth muscle cell lines (ASM cells).

## Data Import

We need two different inputs:

- **countData**: with genes in row and experiments in columns
- **colData**: meta data that describes the column in countData

```{r}
counts <- read.csv("airway_scaledcounts.csv", row.names = 1)
metadata <- read.csv("airway_metadata.csv")
```

A wee peak at counts and meta data

```{r}
head(counts)
head(metadata)
```

> Q1. How many genes are in this dataset?

```{r}
nrow(counts)
```

> Q2. How many 'control' cell lines do we have?

```{r}
sum(metadata$dex == "control")
```

```{r}
table(metadata$dex)
```

## Differential gene expression

We have four replicate drug treated and control (no drug) columns/experiments in our `counts` object

We want one "mean" value for each gene (rows) in "treated" (drug) and one mean value for each gene in "control" columns.

Step 1: Find all "control" columns in `counts`
Step 2: Extract these columns to a new object called `control.counts`
Step 3: Then calculate the mean value for each gene 

Step 1
```{r}
control.inds <- metadata$dex == "control"
```

Step 2
```{r}
control.counts <- counts[, control.inds]
```

Step 3
```{r}
control.mean <- rowMeans(control.counts)
head(control.mean)
```

> Q4. Now do the same thing for the "treated" columns/experiments...

Step 1
```{r}
treated.inds <- metadata$dex == "treated"
```

Step 2
```{r}
treated.counts <- counts[, treated.inds]
```

Step 3
```{r}
treated.mean <- rowMeans(treated.counts)
head(treated.mean)
```


```{r}
meancounts <- data.frame(control.mean, treated.mean)
```

> Q5. Let's create a quick scatterplot showing the mean of the treated samples against the mean of the control samples.

```{r}
library(ggplot2)
ggplot(meancounts) + aes (control.mean, treated.mean) + geom_point(alpha = 0.35) 
```

> Q6. Let's log transform this count data:

```{r}
plot(meancounts, log = "xy")
```

**N.B.** We most often use log2 for this type of data as it makes the interpretation much more straightforward

Treated/Control is often called "fold-change"
```{r}
# If there was no change we would have a log2 fc of 0
log2(10/10)

# If we had double trancript around we would have a log2-fc of 1
log2(20/10)

#If we had half as much trancript around we would have a log2-fc of -1
log2(5/10)
```

> Q. Calculate a log2 fold change value for all our genes and add it as a new column to our `meancounts` object.

```{r}
meancounts$log2fc <- log2(meancounts$treated.mean/
                          meancounts$control.mean)
head(meancounts)
```

There are some "funky" log2fc values (NaN and -Inf here that come about when ever we have 0 mean count values. Typically we would remove these genes from any further analysis - as we can't say anything about them if we have no data for them.

> Q10. Do you trust these results? Why or why not?

We don't trust these results because we've only done our analysis based on fold change, which can be large without being statistically significant (we haven't done anything to determine whether the differences we're seeing are significant). These results in their current form are misleading, so using the DESeq2 package will prove if our results are statistically significant or not.

## DESeq Analysis

Let's do this analysis with an estimate of statistical significance using the **DESeq2** package

```{r}
library(DESeq2)
```

DESeq (like many bioconductor packages) wants its input data in a very specific way.

```{r}
dds <- DESeqDataSetFromMatrix(countData = counts,
                              colData = metadata,
                              design = ~dex)
dds
```

### Run the DESeq analysis pipeline

The main function is called `DESeq()`
```{r}
dds <- DESeq(dds)
```

```{r}
res <- results(dds)
head(res)
```

```{r}
36000 * 0.05
```

## Volcano Plot

This is a main summary results figure from these kinds of studies. It is a plot of Log2-Foldchange vs. (Adjusted) P-value 

```{r}
plot(res$log2FoldChange, 
     res$padj)
```

Again this y-axis is higly needs log transforming and we can flip the y-axis with a minus sign so it looks like every other volcano plot.

```{r}
plot(res$log2FoldChange,
     -log(res$padj))
abline(v = -2, col = "palevioletred")
abline(v = +2, col = "steelblue")
abline(h = -log(0.05), col = "darkseagreen")
```

Start with a default base color "gray"
```{r}
#custom colors
mycols <- rep("gray", nrow(res))
mycols[res$log2FoldChange > 2] <- "steelblue"
mycols[res$log2FoldChange < -2] <- "darkgreen"
mycols[res$padj >= 0.05] <- "grey"

#volcano plot
plot(res$log2FoldChange,
     -log(res$padj),
     col=mycols)

#cut of dash lines
abline(v = c(-2, +2), lty = 2)
abline(h = -log(0.05), lty = 2)
```

> Q. Make a presentation quality ggplot version for this plot. Include clear axis labels, a clean theme, your custom colors, cut-off lines and a plot titles

```{r}
library(ggplot2)

ggplot(res) + 
  aes(log2FoldChange, 
      -log(padj)) + 
  geom_point(colour = mycols) + 
  labs(x = "Log 2 Fold Change",
       y = "-log Adjusted P-Value") +
  geom_vline(xintercept = c(-2,2))+
  geom_hline(yintercept = -log(0.05))+
  theme_bw()
```

## Save our results to a CSV file

```{r}
write.csv(res, file="results.csv")
```


## Add annotation data

We need to add missing annotation data to our main `res` results object. This includes the common gene "symbol"

```{r}
head(res)
```

We will use R and bioconductor to do this "ID mapping"

```{r}
library("AnnotationDbi")
library("org.Hs.eg.db")
```

Let's see what databases we can use for translation/mapping...
```{r}
columns(org.Hs.eg.db)
```

We can use the `mapIds()` function now to "translate" between any of these databases.

```{r}
res$symbol <- mapIds(org.Hs.eg.db,
              keys=row.names(res), # Our genenames
              keytype="ENSEMBL",   # The format of our genenames
              column="SYMBOL")     # The new format we want to add
```

> Q. Also add "ENTREZID", "GENENAME"

```{r}
res$entrez <- mapIds(org.Hs.eg.db,
              keys=row.names(res), # Our genenames
              keytype="ENSEMBL",   # The format of our genenames
              column="ENTREZID")   # The new format we want to add
  
res$genename <- mapIds(org.Hs.eg.db,
                keys=row.names(res), # Our genenames
                keytype="ENSEMBL",   # The format of our genenames
                column="GENENAME")   # The new format we want to add
```

```{r}
head(res)
```

## Save annotated results to a CSV file

```{r}
write.csv(res, file = "results_annotated.csv")
```

## Pathway Analysis

What known biological pathways do our differentially expressed genes overlap with (i.e. play a role in)?

There's lots of bioconductor packages to do this type of analysis.

We will use one of the oldest called **gage** along with **pathview** to render nice pics of the pathways we find.

`BiocManager::install(c("pathview","gage","gageData"))`

```{r, message = FALSE}
library(pathview)
library(gage)
library(gageData)
```
Have a wee peak what is in `gageData`

```{r}
# Examine the first 2 pathways in this kegg set for humans
data(kegg.sets.hs)
head(kegg.sets.hs, 2)
```

The main `gage()` function that does the work wants a simple vector as input.

```{r}
foldchanges <-res$log2FoldChange
names(foldchanges) <- res$symbol
head(foldchanges)
```

The KEGG database uses ENTREZ ids so we need to provide these in our input vector for **gage**:
```{r}
names(foldchanges) <- res$entrez
```

Now we can run `gage()`

```{r}
# Get the results
keggres = gage(foldchanges, gsets=kegg.sets.hs)
```

What is in the output object `keggres`

```{r}
attributes(keggres)
```

```{r}
# Look at the first three down (less) pathways
head(keggres$less, 3)
```

We can use the **pathview** function to render a figure of any of these pathways along with annotation for our DEGs

Let's see the hsa05310 Asthma pathway with our DEGs colored up:

```{r}
pathview(gene.data=foldchanges, pathway.id="hsa05310")
```
![](hsa05310.pathview.png)


> Q. Can you render and insert here the pathway figure for "Graft-versus-host-disease" and "Type I diabetes"

```{r}
pathview(gene.data=foldchanges, pathway.id="hsa05332")
pathview(gene.data=foldchanges, pathway.id="hsa04940")
```
![](hsa05332.pathview.png)

![](hsa04940.pathview.png)