---
title: "Class 11: AlphaFold"
author: "Katherine Quach (A18541014)"
format: pdf
toc: TRUE
---

## Background

We saw last day that the main repository for biomolecular structure (the PDB database) only has ~250,000 entries.

UniProtKB (the main protein sequence database) has over 200 million entries!

In this hands-on session we will utilize AlphaFold to predict protein structure from sequence (Jumper et al. 2021).

Without the aid of such approaches, it can take years of expensive laboratory work to determine the structure of just one protein. With AlphaFold we can now accurately compute a typical protein structure in as little as ten minutes.

## The EBI AlphaFold Database

The EBI AlphaFold database contains lots of computed structure models. It is increasingly likely that the structure you are interested in is already in this database <https://alphafold.ebi.ac.uk/>

There are 3 major outputs from AlphaFold

1. A model of the structure in **PDB format**
2. A **pLDDT score** that tells us how confident the model is for a given residue in you proteins (high values are good, above 70)
3. A **PAE score** that tells us about protein packing quality

If you can't find a matching entry for the sequence you are interested in AFDB you can run AlphaFold yourself...

## Running AlphaFold

We will use ColabFold to run AlphaFold on our sequence <https://github.com/sokrypton/ColabFold?tab=readme-ov-file>


Figure from AlphaFold here!

## Interpreting Results

Custom analysis of resulting models

We can read all the AlphaFold results into R and do more quantitative analysis than just viewing the structures in Mol-star:

```{r}
library(bio3d)
p <- read.pdb("hivpr_23119/")
```


```{r}
pdb_files <- list.files("hivpr_23119/", pattern = "*.pdb", full.names = T)
basename(pdb_files)
```

```{r}
library(bio3dview)
pdbs <- pdbaln(pdb_files, fit = TRUE, exefile = "msa")

view.pdbs(pdbs)

# Quick view of model sequence
pdbs
```

```{r}
rd <- rmsd(pdbs, fit = T)
range(rd)

library(pheatmap)

colnames(rd) <- paste0("m", 1:5)
rownames(rd) <- paste0("m", 1:5)
pheatmap(rd)
```

```{r}
pdb <- read.pdb("1hsg")
```

```{r}
plotb3(pdbs$b[1,], typ="l", lwd=2, sse=pdb)
points(pdbs$b[2,], typ="l", col="red")
points(pdbs$b[3,], typ="l", col="blue")
points(pdbs$b[4,], typ="l", col="darkgreen")
points(pdbs$b[5,], typ="l", col="orange")
abline(v=100, col="gray")

core <- core.find(pdbs)

core.inds <- print(core, vol=0.5)

xyz <- pdbfit(pdbs, core.inds, outpath="corefit_structures")
```

```{r}
rf <- rmsf(xyz)

plotb3(rf, sse=pdb)
abline(v=100, col="gray", ylab="RMSF")
```

## Predicted Alignment Error for Domains

```{r}
library(jsonlite)

# Listing of all PAE JSON files
pae_files <- list.files(path="hivpr_23119/",
                        pattern=".*model.*\\.json",
                        full.names = TRUE)

# Read the 1st and 5th files
pae1 <- read_json(pae_files[1],simplifyVector = TRUE)
pae2 <- read_json(pae_files[2],simplifyVector = TRUE)
pae3 <- read_json(pae_files[3],simplifyVector = TRUE)
pae4 <- read_json(pae_files[4],simplifyVector = TRUE)
pae5 <- read_json(pae_files[5],simplifyVector = TRUE)

attributes(pae1)

# Per-residue pLDDT scores (same as B-factorof PDB)
head(pae1$plddt)
head(pae2$plddt)
head(pae3$plddt)
head(pae4$plddt)
head(pae5$plddt)

# Use max PAE values for ranking models
pae1$max_pae
pae2$max_pae
pae3$max_pae
pae4$max_pae
pae5$max_pae
```

```{r}
# Plot N by N PAE scores with ggplot or functions from Bio3D package
plot.dmat(pae1$pae, 
          xlab="Residue Position (i)",
          ylab="Residue Position (j)",  
          grid.col = "black",
          zlim = c(0,30))

plot.dmat(pae2$pae, 
          xlab="Residue Position (i)",
          ylab="Residue Position (j)",  
          grid.col = "black",
          zlim = c(0,30))

plot.dmat(pae3$pae, 
          xlab="Residue Position (i)",
          ylab="Residue Position (j)",  
          grid.col = "black",
          zlim = c(0,30))

plot.dmat(pae4$pae, 
          xlab="Residue Position (i)",
          ylab="Residue Position (j)",  
          grid.col = "black",
          zlim = c(0,30))

plot.dmat(pae5$pae, 
          xlab="Residue Position (i)",
          ylab="Residue Position (j)",  
          grid.col = "black",
          zlim = c(0,30))
```

## Residue Conservation from Alignment File

```{r}
aln_file <- list.files(path="hivpr_23119/",
                       pattern=".a3m$",
                        full.names = TRUE)
aln_file

aln <- read.fasta(aln_file[1], to.upper = TRUE)

# How many sequences are in this alignment?
dim(aln$ali)

# Score residue conservation in alignment
sim <- conserv(aln)
plotb3(sim[1:99], sse=trim.pdb(pdb, chain="A"),
       ylab="Conservation Score")

con <- consensus(aln, cutoff = 0.9)
con$seq

m1.pdb <- read.pdb(pdb_files[1])
occ <- vec2resno(c(sim[1:99], sim[1:99]), m1.pdb$atom$resno)
write.pdb(m1.pdb, o=occ, file="m1_conserv.pdb")
```